### PR TITLE
Retroachievements support (work-in-progress)

### PR DESCRIPTION
**Pull request by hrydgard**
Implement RetroAchievements support (#16169). Not to be confused with RetroArch, these are two different projects.

Most of RetroAchievements.cpp was adapted from [DuckStation's integration](https://github.com/stenzek/duckstation/blob/master/src/frontend-common/achievements.cpp) by @stenzek.

So far, login works (unless you're on a touchscreen device), game identification and basic achievement listing works, and unlocking achivements works.

TODO list before initial merge:

- [x] Login UI, basic settings
- [x] Proper achievement listing UI from the pause screen
- [x] On-screen notification UI for unlocks
- [x] Fix the game hashing so it can work with scoped storage
- [x] More secure login token storage
- [x] Fill in the remaining translation strings in en_US.ini (will add actual translations in the next PR)

Next steps after this:

* Leaderboards UI
* HTTPS support (we force HTTP because our client implemenetation lacks ssl support)
* Challenge mode (disable savestates etc)
* Juicy sound effect for unlocks
* Other missing features

### CODE REVIEW

**Comment by cannnAvar** [comment]
How we can use achivements now

**Comment by hrydgard** [comment]
Stay tuned, it's not ready enough yet.

**Comment by hrydgard** [comment]
Rebased on top of #17594 .

**Comment by stenzek** [comment]
Got pinged about this, please maintain my copyright notice in DuckStation-derived files (and add your own, obviously), thanks.

Also, feel free to relicense those files to GPL2 to match the rest of the emu if that makes things easier (in DS it's GPL3).

**Comment by hrydgard** [comment]
@stenzek will do, thank you!

**Comment by ScottCarammell** [comment]
> 

If you wanna get PSP achievements right now, you can use the PPSSPP core in RetroArch. It's not optimal, but it'll be the only option until this is finished.

**Comment by hrydgard** [comment]
Progress report: You can now login through the UI, and play and get achievements, and you do get nice little popups with achievement icons and text when you do. Logging in works on all desktop platforms currently - or if you have a physical keyboard for your phone somehow. This will of course be fixed soon.

I'll do a more general fix for text input UIs on master soon, then bring it into this PR.

**Comment by hrydgard** [comment]
This is starting to get usable now, opening for review, but don't merge until the below is solved:

I think we should probably store the RetroAchievements login token somewhere slightly more secure than in PPSSPP.ini, it's too easy to accidentally leak your token while sending that file in for debugging or something. 

On Android we'll just store it in a file in the app private directory.

On Windows, Mac, Linux, I'm thinking just store it in a separate file in the home directory somewhere?

**Comment by Captainquill** [comment]
Hello sir.. When will this be ready to download. Just a tentative release Window.?

**Comment by hrydgard** [comment]
Next week it will be in the "nightly" builds.

**Comment by ANR2ME** [comment]
> Hello sir.. When will this be ready to download. Just a tentative release Window.?

Currently you can also download the artifacts from https://github.com/hrydgard/ppsspp/actions/runs/5360144898?pr=17589 if you want to test it.

**Comment by hrydgard** [comment]
Should be ready for initial review/merge now.

**Comment by hrydgard** [comment]
Hm, actually need to clean up the choice of directory for secrets storage.

**Comment by hrydgard** [comment]
Ok, I've been thinking about where to put the login token on various platforms. One idea would be the following:

Android: It clearly belongs in the app private data folder. 
Windows: There are two cases here:
  * installed: %APPDATA%\PPSSPP ?
  * portable: memstick/PSP/SYSTEM , next to the ini (but not included in the ini file)
Linux/Mac: memstick/PSP/SYSTEM
iOS: ?
Switch: ?

However, after some chats I think I might be overthinking it. It's only really important to keep it out of the .ini as we often ask people to post that. But just putting the login token in a file next to the ini on all platforms may be good enough, really.

**Comment by hrydgard** [comment]
Alright, I think this is a pretty good checkpoint, and I'm just gonna get it in and work from that state.

File: Common/UI/IconCache.h [hunk 1]
```diff
@@ -51,6 +51,16 @@ class IconCache {
 
 	IconCacheStats GetStats();
 
+	// for testing
+	std::string GetFirstIconName() const {
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply Common/UI/IconCache.h:55]: I never used it for anything, I'll just delete it.

**Pull request merged**